backend/tracim_backend/lib/utils/authentication.py

```diff
@@ -20,16 +28,195 @@
+
+        user_api = UserApi(None, request.dbsession, self.app_config)
+
+        user = user_api.saml_authenticate(
+            user=None,
+            user_id=saml_user_id,
+            name=request.session.get("saml_name"),
+            mail=request.session.get("saml_mail"),
+        )
+        self.remember(request, user.user_id)
+        request.set_user(user)
+        return user.user_id
+
+    def permits(self, request, context, permission):
+        # TODO - M.L - 2023/09/06 - Maybe check for session lifetime or validity
+        # if request.authenticated_userid is None \
+        #     and not request.path.startswith("/saml") \
+        #     and not request.path.startswith("/saml") \
+        #     and not request.path is "/ui/login":
+        #     return Denied("Nobody is allowed")
+        return Allowed("Allowed")
+
+    def remember(
+        self, request: TracimRequest, userid: typing.Union[str, int], **kw: typing.Any
+    ) -> typing.Iterable[str]:
+        self._session_helper.remember(request, userid)
+        return []
+
+    def forget(self, request: TracimRequest, **kw: typing.Any) -> typing.Iterable[str]:
+        if "saml_user_id" in request.session:
+            del request.session["saml_user_id"]
+            del request.session["saml_mail"]
+            del request.session["saml_name"]
+        self._session_helper.forget(request, **kw)
+        return []
+
+    def _idp_chooser(self, request: TracimRequest) -> Response:
+        # TODO - SGD - 2023-07-26 - This view should render an HTML form
+        # with the list of configured IdPs.
+        # The form should redirect to the chosen IdPs SSO endpoint when submitted.
+        # Maybe some JS will be needed here?
+        return Response(
+            f"""
+          <form action="{self.app_config.WEBSITE__BASE_URL}/saml/acs" method="POST">
+            <input name="saml_user_id" type="hidden" value="saml-user"/>
+            <input name="redirect_url" type="hidden" value="{request.path_url}" />
+            <button>Login</button>
+          </form>"""
+        )
+
+    def _acs(self, request: TracimRequest) -> Response:
+        response = request.POST['SAMLResponse']
+        try:
+            authn_response = self.saml_client.parse_authn_request_response(
+                response,
+                binding=BINDING_HTTP_POST,
+            )
+        except ResponseLifetimeExceed as e:
+            return Response(e.__str__())
+        identity = authn_response.get_identity()
+        if identity is None or "UserID" not in identity:
+            return HTTPBadRequest()
+
+        # TODO - M.L - 2023/09/06 - Associate response / expiry date to session
+        request.session["saml_user_id"] = ''.join(identity["UserID"])
+        request.session["saml_mail"] = ''.join(identity["EmailAddress"])
+        request.session[
+            "saml_name"] = f'{"".join(identity["FirstName"])} {"".join(identity["LastName"])}'
+
+        return HTTPFound("/")
```

Comment:
This functions parses the response from the IdP (well, technically from the users, as the response is encoded in the POST data but well) does some little checks (is the IdP known, is the response expired, etc...) and matches hardcoded attributes to session storage. Authentication checks are then done with the `authenticated_userid` function